### ISSUE TITLE
Finagle exception: “This writer’s reader has been discarded”

### ISSUE THREAD

**Issue by prdoyle** [linkerd/linkerd#1599]
Issue Type:
- [x] Bug report

**What happened**:

I brought up a simple linkerd+namerd configuration using docker-compose, with three containers: linkerd, namerd, and my service.  Ran a workload hitting my API about 30 times per second.  My service is configured so that each of these API calls results in two requests to linkerd, which is configured to route those back to another API on the same instance of the same service.  (My service is effectively recursing via linkerd.)

Despite all my APIs returning 200s, a few times per minute, I got `502: Bad Gateway` from linkerd, along with this exception backtrace on the console:

```
W 0817 15:16:33.020 UTC THREAD24 TraceId:6cf8fbc2d0b9399f: Exception propagated to the default monitor (upstream address: /172.20.0.3:60158, downstream address: /172.20.0.3:4567, label: #/io.l5d.fs/icecube).

com.twitter.io.Reader$ReaderDiscarded: This writer's reader has been discarded
   at com.twitter.finagle.netty4.http.StreamTransports$$anon$1.discard(StreamTransports.scala:70)
   at com.twitter.finagle.http.DelayedReleaseService$$anon$2.discard(DelayedReleaseService.scala:56)
   at com.twitter.finagle.http.codec.HttpServerDispatcher$$anonfun$handle$2.applyOrElse(HttpServerDispatcher.scala:61)
   at com.twitter.finagle.http.codec.HttpServerDispatcher$$anonfun$handle$2.applyOrElse(HttpServerDispatcher.scala:60)
   at com.twitter.util.Promise.raise(Promise.scala:681)
   at com.twitter.util.Future$JoinPromise$$anonfun$2.applyOrElse(Future.scala:254)
   at com.twitter.util.Future$JoinPromise$$anonfun$2.applyOrElse(Future.scala:251)
   at com.twitter.util.Promise.raise(Promise.scala:681)
   at com.twitter.util.Promise.raise(Promise.scala:686)
   at com.twitter.finagle.http.exp.GenSerialServerDispatcher.$anonfun$new$1(ServerDispatcher.scala:125)
   at com.twitter.util.Future.$anonfun$ensure$1(Future.scala:941)
   at com.twitter.util.Future.$anonfun$ensure$1$adapted(Future.scala:941)
   at com.twitter.util.Promise$Monitored.apply(Promise.scala:202)
   at com.twitter.util.Promise$Monitored.apply(Promise.scala:193)
   at com.twitter.util.Promise$$anon$7.run(Promise.scala:530)
   at com.twitter.concurrent.LocalScheduler$Activation.run(Scheduler.scala:200)
   at com.twitter.concurrent.LocalScheduler$Activation.submit(Scheduler.scala:158)
   at com.twitter.concurrent.LocalScheduler.submit(Scheduler.scala:272)
   at com.twitter.concurrent.Scheduler$.submit(Scheduler.scala:108)
   at com.twitter.util.Promise.runq(Promise.scala:520)
   at com.twitter.util.Promise.updateIfEmpty(Promise.scala:877)
   at com.twitter.finagle.netty4.transport.ChannelTransport.com$twitter$finagle$netty4$transport$ChannelTransport$$fail(ChannelTransport.scala:95)
   at com.twitter.finagle.netty4.transport.ChannelTransport$$anon$1.channelInactive(ChannelTransport.scala:186)
   at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
```

**What you expected to happen**:

Since my service returned 200 for all requests, I expected linkerd to do the same.

**How to reproduce it (as minimally and precisely as possible)**:

Use docker-compose to run three containers:
- linkerd
- namerd
- An http service which, when one API is called, invokes two other APIs on the same service via linkerd

**Anything else we need to know?**:
*Some background:* I'm attempting to use linkerd and namerd to manage shards in an in-memory database service I'm developing.  A query request hits my service, which determines that the queried database has two shards.  It therefore produces two more API calls to get the shards, and those requests are routed via linkerd to a node that houses each shard.  For now, I only have one node running my service, so the shards happen to be on the same node; therefore, linkerd routes the shard requests back to the same host, effectively causing my service to recurse via linkerd.

**Environment**:
- linkerd and namerd 1.1.3 from Docker Hub
- Windows 10 plus Docker for Windows
- My Thinkpad

My service is written in Java using the Java Spark framework, and Java 8 streams to make the shard requests run in parallel.

My config files, plus linkerd metrics, are in this Discourse article:

https://discourse.linkerd.io/t/finagle-exception-this-writers-reader-has-been-discarded/244

**Comment by prdoyle** [comment]
Tried with linkerd+namerd 1.1.0, same problem.

**Comment by wmorgan** [comment]
Thanks @prdoyle. We'll dig into this.

**Comment by hawkw** [comment]
Hi @prdoyle, I strongly suspect PR #1609 will fix this issue as well, since it adds handling for the exception that's being thrown here. 

When we've tried to reproduce this issue, we don't see the 502 responses you're describing. Since I don't have your HTTP service to test against, can I get you to confirm whether that PR fixes this issue?  
+ Check out the [`eliza/fix-spurious-exception`](https://github.com/linkerd/linkerd/tree/eliza/fix-spurious-exception) branch 
+ Build a Linkerd docker image by running `./sbt docker` from the root dir of the Linkerd repo. The image created will be tagged `1.1.3-SNAPSHOT`.
+ Change your `docker-compose.yml` to use the `buoyantio:linkerd:1.1.3-SNAPSHOT` image for Linkerd 
+ Run `docker-compose up` and check whether this issue persists.

**Comment by hawkw** [comment]
Quick update @prdoyle: we've recently cut a release candidate build of 1.2.0 for canarying some Kubernetes changes, and you should be able to test using that rather than building a new docker image. Just change your` docker-compose.yml` to use the `buoyantio:linkerd:1.2.0-rc1` image, & let me know if this problem is still occurring?

**Comment by prdoyle** [comment]
Thanks @hawkw, tried it and still got the failure.

```
I 0826 20:28:31.207 UTC THREAD1: linkerd 1.2.0-rc1 (rev=a044149e51936068a4b255f0acf30f582bb077bf) built at 20170825-135459
```
...
```
com.twitter.io.Reader$ReaderDiscarded: This writer's reader has been discarded
   at com.twitter.finagle.netty4.http.StreamTransports$$anon$1.discard(StreamTransports.scala:70)
   at com.twitter.finagle.http.DelayedReleaseService$$anon$2.discard(DelayedReleaseService.scala:56)
   at com.twitter.finagle.http.codec.HttpServerDispatcher$$anonfun$handle$2.applyOrElse(HttpServerDispatcher.scala:61)
   at com.twitter.finagle.http.codec.HttpServerDispatcher$$anonfun$handle$2.applyOrElse(HttpServerDispatcher.scala:60)
   at com.twitter.util.Promise.raise(Promise.scala:681)
   at com.twitter.util.Future$JoinPromise$$anonfun$2.applyOrElse(Future.scala:254)
   at com.twitter.util.Future$JoinPromise$$anonfun$2.applyOrElse(Future.scala:251)
   at com.twitter.util.Promise.raise(Promise.scala:681)
   at com.twitter.util.Promise.raise(Promise.scala:686)
   at com.twitter.finagle.http.exp.GenSerialServerDispatcher.$anonfun$new$1(ServerDispatcher.scala:125)
   at com.twitter.util.Future.$anonfun$ensure$1(Future.scala:941)
   at com.twitter.util.Future.$anonfun$ensure$1$adapted(Future.scala:941)
   at com.twitter.util.Promise$Monitored.apply(Promise.scala:202)
   at com.twitter.util.Promise$Monitored.apply(Promise.scala:193)
   at com.twitter.util.Promise$$anon$7.run(Promise.scala:530)
   at com.twitter.concurrent.LocalScheduler$Activation.run(Scheduler.scala:200)
   at com.twitter.concurrent.LocalScheduler$Activation.submit(Scheduler.scala:158)
   at com.twitter.concurrent.LocalScheduler.submit(Scheduler.scala:272)
   at com.twitter.concurrent.Scheduler$.submit(Scheduler.scala:108)
   at com.twitter.util.Promise.runq(Promise.scala:520)
   at com.twitter.util.Promise.updateIfEmpty(Promise.scala:877)
   at com.twitter.finagle.netty4.transport.ChannelTransport.com$twitter$finagle$netty4$transport$ChannelTransport$$fail(ChannelTransport.scala:95)
   at com.twitter.finagle.netty4.transport.ChannelTransport$$anon$1.channelInactive(ChannelTransport.scala:186)
   at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
   at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
   at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
   at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
```

**Comment by hawkw** [comment]
@prdoyle okay, thanks – I'll continue looking into this issue, then. Are you still getting 502 responses?

**Comment by prdoyle** [comment]
@hawkw - Yep, still 502 Bad Gateway.

**Comment by prdoyle** [comment]
When I get some time, I'm hoping to reduce the testcase to a few Dockerfiles and a docker-compose.yml for you.  So, the more you procrastinate, the easier this may become to reproduce...

**Comment by hawkw** [comment]
Thanks @prdoyle, that will make diagnosing this issue a lot easier! I haven't yet been able to reproduce this failure, and I suspect the bug is only revealed by the particular behavior of your service.

**Comment by adleong** [comment]
Hi @prdoyle, we still haven't been able to repro this internally.  Any additional repro instructions would be very helpful.

**Comment by d10i** [comment]
I'm also seeing this on both 1.1.2 and 1.2.1. Tried to reproduce but haven't been able to...

**Comment by adleong** [comment]
We can't really make any progress here unless we are able to repro.

**Comment by d10i** [comment]
Yeah I know, just wanted to pointed out https://github.com/linkerd/linkerd/pull/1609 definitely didn't fix the problem as it's still happening on 1.2.1.

It seems to only happen when we fire > 5 requests in parallel to the same service and same endpoint (with different payload). We get HTTP 502 with `This writer's reader has been discarded` in the response body. Only one of the requests that are fired in parallel fails and it doesn't happen all the time (happened 12 times out of 68 in the past 48 hours). I suspect race condition...

If I have time I'll try to setup a docker-compose environment that reproduces the issue.

**Comment by d10i** [comment]
Hi @adleong and @hawkw, I finally managed to create a project to reproduce this issue. You can find it here: https://github.com/Attest/writers-reader-discarded. The README explains how it works and how to run it but let me know if you need help!

**Comment by wmorgan** [comment]
@d10i that's awesome. nice work. we'll dig into this.

**Comment by d10i** [comment]
FYI I've just tried with linkerd 1.3.2 and it doesn't fail. It must have been fixed as a side effect of another change?

**Comment by hawkw** [comment]
@d10i hm, that's interesting – off the top of my head, I can't think of any changes in 1.3.2 that would fix this issue, but we can take a closer look. @prdoyle, can you confirm whether or not you're seeing this error on 1.3.2?

**Comment by d10i** [comment]
1.3.0 fixes it too and by looking at https://github.com/linkerd/linkerd/compare/1.2.1...1.3.0 I think it's the upgrade to finagle 7 that might have fixed it?

**Comment by hawkw** [comment]
@d10i that certainly makes sense; when I was looking into this issue a couple weeks ago, it seemed like it was happening somewhere in Finagle.  I didn't realise this was fixed on 1.3.0 as well, but I suspect you're right that the Finagle 7 upgrade fixed this.

@prdoyle I'm going to go ahead and close this issue; if you still see this error on a version >1.3.0, please feel free to reopen.

**Comment by prdoyle** [comment]
@hawkw sounds good.  I'll reopen if I see it again.  Thanks!
